packages/map/src/index.ts

```diff
@@ -21,100 +21,114 @@ const $KEYS = Symbol("track-keys");
-    super();
-    if (initial) for (const v of initial) super.set(v[0], v[1]);
+  [Symbol.iterator](): MapIterator<[K, V]> {
+    return this.entries();
   }
 
-  // reads
-  has(key: K): boolean {
-    this.#keyTriggers.track(key);
-    return super.has(key);
-  }
-  get(key: K): V | undefined {
-    this.#valueTriggers.track(key);
-    return super.get(key);
+  constructor(entries?: Iterable<readonly [K, V]> | null) {
+    super();
+    if (entries) for (const entry of entries) super.set(...entry);
   }
+
   get size(): number {
-    this.#keyTriggers.track($KEYS);
+    this.#keyTriggers.track($OBJECT);
     return super.size;
   }
 
   *keys(): MapIterator<K> {
+    this.#keyTriggers.track($OBJECT);
+
     for (const key of super.keys()) {
-      this.#keyTriggers.track(key);
       yield key;
     }
-    this.#keyTriggers.track($KEYS);
   }
+
   *values(): MapIterator<V> {
-    for (const [key, v] of super.entries()) {
-      this.#valueTriggers.track(key);
-      yield v;
+    this.#valueTriggers.track($OBJECT);
+
+    for (const value of super.values()) {
+      yield value;
     }
-    this.#keyTriggers.track($KEYS);
   }
+
   *entries(): MapIterator<[K, V]> {
+    this.#keyTriggers.track($OBJECT);
+    this.#valueTriggers.track($OBJECT);
+
     for (const entry of super.entries()) {
-      this.#valueTriggers.track(entry[0]);
       yield entry;
     }
-    this.#keyTriggers.track($KEYS);
   }
 
-  // writes
+  forEach(callbackfn: (value: V, key: K, map: Map<K, V>) => void, thisArg?: any): void {
+    this.#keyTriggers.track($OBJECT);
+    this.#valueTriggers.track($OBJECT);
+    super.forEach(callbackfn, thisArg);
+  }
+
+  has(key: K): boolean {
+    this.#keyTriggers.track(key);
+    return super.has(key);
+  }
+
+  get(key: K): V | undefined {
+    this.#valueTriggers.track(key);
+    return super.get(key);
+  }
+
   set(key: K, value: V): this {
-    batch(() => {
-      if (super.has(key)) {
-        if (super.get(key)! === value) return;
-      } else {
-        this.#keyTriggers.dirty(key);
-        this.#keyTriggers.dirty($KEYS);
-      }
-      this.#valueTriggers.dirty(key);
-      super.set(key, value);
-    });
-    return this;
+    const hadNoKey = !super.has(key);
+    const hasChanged = super.get(key) !== value;
+    const result = super.set(key, value);
+
+    if (hasChanged || hadNoKey) {
+      batch(() => {
+        if (hadNoKey) {
+          this.#keyTriggers.dirty($OBJECT);
+          this.#keyTriggers.dirty(key);
+        }
+        if (hasChanged) {
+          this.#valueTriggers.dirty($OBJECT);
+          this.#valueTriggers.dirty(key);
+        }
+      });
+    }
+
+    return result;
   }
+
   delete(key: K): boolean {
-    const r = super.delete(key);
-    if (r) {
+    const isDefined = super.get(key) !== undefined;
+    const result = super.delete(key);
+
+    if (result) {
       batch(() => {
+        this.#keyTriggers.dirty($OBJECT);
+        this.#valueTriggers.dirty($OBJECT);
```

Comment:
This is important: we do dirty global values trigger **always**, as values() may depend on order. Even though the actual value of the key may not change, removing it, will have impact on amount of iterations in *values()